mesonbuild/build.py

```diff
@@ -1688,6 +1688,26 @@ def check_module_linking(self):
                                      'use shared_library() with `override_options: [\'b_lundef=false\']` instead.')
                     link_target.force_soname = True
 
+    def process_vs_module_defs_kw(self, kwargs: T.Dict[str, T.Any]) -> None:
+        if kwargs.get('vs_module_defs') is not None:
+            path: T.Union[str, File, CustomTarget, CustomTargetIndex] = kwargs['vs_module_defs']
+            if isinstance(path, str):
+                if os.path.isabs(path):
+                    self.vs_module_defs = File.from_absolute_file(path)
+                else:
+                    self.vs_module_defs = File.from_source_file(self.environment.source_dir, self.subdir, path)
+            elif isinstance(path, File):
+                # When passing a generated file.
+                self.vs_module_defs = path
+            elif isinstance(path, (CustomTarget, CustomTargetIndex)):
+                # When passing output of a Custom Target
+                self.vs_module_defs = File.from_built_file(path.get_subdir(), path.get_filename())
+            else:
+                raise InvalidArguments(
+                    'Shared library vs_module_defs must be either a string, '
+                    'a file object or a Custom Target')
```

Comment:
This can also be a CustomTargetIndex now too right?

Reply:
Yup. Probably also shouldn't say "shared library" once it's added to vs_module_defs either. It will sure be nice to get rid of all of this build layer validation duplicaiton

Reply:
I agree